src/libs/ck-libs/channel/channel.h

```diff
@@ -0,0 +1,180 @@
+  };
+  std::array<std::queue<std::shared_ptr<CkWaitRequest>>, N> waiting;
+public:
+  // Returns true when a thread is waiting on a channel
+  // Stores the waiting thread in the reference argument
+  bool hasWaiter(int channel, CthThread &thread) {
+    if (waiting[channel].empty()) {
+      return false;
+    } else {
+      // Pop a request from the wait-queue
+      auto request = waiting[channel].front();
+      waiting[channel].pop();
+      // If it has expired
+      if (request->expired) {
+        // Recurse and check for another waiter
+        return hasWaiter(channel, thread);
+      } else {
+        // Otherwise, mark it as expired
+        request->expired = true;
+        // And set the return values
+        thread = request->thread;
+        return true;
+      }
+    }
+  }
+  // Places a waiter into the queue for the specified channel
+  void wait(int channel, const CthThread &thread) {
+    waiting[channel].emplace(new CkWaitRequest(thread));
+  }
+  // Places a (common) waiter into the queue for all specified channels
+  void waitForAny(int* channels, int numChannels, const CthThread &thread) {
+    std::shared_ptr<CkWaitRequest> ptr(new CkWaitRequest(thread));
+    for (int i = 0; i < numChannels; i++) {
+      waiting[channels[i]].push(ptr);
+    }
+  }
+};
+
+template<int N, typename T>
+class CkMultiChannel : public CBase_CkMultiChannel<N, T> {
+  template<typename Value>
+  inline void sendToCallback(CkCallback cb, Value value) {
+    PUP::sizer szr;
+    szr | value;
+    CkMarshallMsg *msg = CkAllocateMarshallMsg(szr.size(), NULL);
+    PUP::toMem ppr((void *)msg->msgBuf);
+    ppr | value;
+    cb.send(msg);
+  }
+public:
+  CkMultiChannel() {}
+  CkMultiChannel(CkMigrateMessage*) {}
+
+  void send(int channel, T value) {
+    channel = channel % N;
+    data[channel].push(value);
+    CthThread thread;
+    if (waiting.hasWaiter(channel, thread)) {
+      CthAwaken(thread);
+    }
+  }
+
+  T receive(int channel) {
+    channel = channel % N;
+    if (data[channel].empty()) {
+      waiting.wait(channel, CthSelf());
+      CthSuspend();
+    }
+    auto front = data[channel].front();
+    data[channel].pop();
+    return front;
+  }
+
+  void receive(int channel, CkCallback cb) {
+    sendToCallback(cb, this->receive(channel));
+  }
+
+  void waitAll(int* channels, int numChannels) {
+    for (int i = 0; i < numChannels; i++) {
+      this->receive(channels[i]);
+    }
+  }
+
+  void waitN(int channel, int numMsgs) {
+    for (int i = 0; i < numMsgs; i++) {
+      this->receive(channel);
+    }
+  }
+
+  int checkChannels(int* channels, int numChannels) {
+    for (int i = 0; i < numChannels; i++) {
+      int channel = channels[i] % N;
+      if (!data[channel].empty()) {
+        return channel;
+      }
+    }
+    return -1;
+  }
+
+  std::pair<int, T> receiveFromAny(int* channels, int numChannels) {
+    int which = checkChannels(channels, numChannels);
+    if (which >= 0) goto END;
```

Comment:
Probably nicer to change this to `if (which < 0) { ... }` and get rid of the goto

Reply:
Alright, I put the GOTO there to avoid repeating the following block of code, but if that's not a concern then I'll duplicate it and delete the GOTO.

    auto front = data[which].front();
    data[which].pop();
    return std::make_pair(which, front);

Reply:
Isn't this equivalent?
```
std::pair<int, T> receiveFromAny(int* channels, int numChannels) {
    int which = checkChannels(channels, numChannels);
    if (which < 0) {
       waiting.waitForAny(channels, numChannels, CthSelf());
       CthSuspend();
       which = checkChannels(channels, numChannels);
    }
    auto front = data[which].front();
    data[which].pop();
    return std::make_pair(which, front);
  }
```